src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -688,67 +918,187 @@ void ShenandoahFreeSet::rebuild() {
+        assert(!is_old_collector_free(idx), "We are about to add it, it shouldn't be there already");
+        set_old_collector_free(idx);
+        log_debug(gc)("  Setting Region " SIZE_FORMAT " _old_collector_free_bitmap bit to true", idx);
+      } else {
+        _capacity += alloc_capacity(region);
 
-      log_debug(gc, free)("  Adding Region " SIZE_FORMAT " (Free: " SIZE_FORMAT "%s, Used: " SIZE_FORMAT "%s) to mutator free set",
+        assert(_used <= _capacity, "must not use more than we have");
+        assert(!is_mutator_free(idx), "We are about to add it, it shouldn't be there already");
+        set_mutator_free(idx);
+        log_debug(gc, free)(
+          "  Adding Region " SIZE_FORMAT " (Free: " SIZE_FORMAT "%s, Used: " SIZE_FORMAT "%s) to mutator free set",
           idx, byte_size_in_proper_unit(region->free()), proper_unit_for_byte_size(region->free()),
-               byte_size_in_proper_unit(region->used()), proper_unit_for_byte_size(region->used()));
+          byte_size_in_proper_unit(region->used()), proper_unit_for_byte_size(region->used()));
+      }
     }
   }
 
   // Evac reserve: reserve trailing space for evacuations
+  size_t young_reserve, old_reserve;
   if (!_heap->mode()->is_generational()) {
-    size_t to_reserve = (_heap->max_capacity() / 100) * ShenandoahEvacReserve;
-    reserve_regions(to_reserve);
+    young_reserve = (_heap->max_capacity() / 100) * ShenandoahEvacReserve;
+    old_reserve = 0;
   } else {
-    size_t young_reserve = (_heap->young_generation()->max_capacity() / 100) * ShenandoahEvacReserve;
+
     // Note that all allocations performed from old-gen are performed by GC, generally using PLABs for both
     // promotions and evacuations.  The partition between which old memory is reserved for evacuation and
     // which is reserved for promotion is enforced using thread-local variables that prescribe intentons within
     // each PLAB.  We do not reserve any of old-gen memory in order to facilitate the loaning of old-gen memory
     // to young-gen purposes.
-    size_t old_reserve = 0;
-    size_t to_reserve = young_reserve + old_reserve;
-    reserve_regions(to_reserve);
+    if (_heap->has_evacuation_reserve_quantities()) {
+      // We are rebuilding at the end of final mark, having established evacuation budgets for this GC pass.
+      young_reserve = _heap->get_young_evac_reserve();
+      old_reserve = _heap->get_promoted_reserve() + _heap->get_old_evac_reserve();
+    } else {
+      // We are rebuilding at end of GC, so we set aside budgets specified on command line (or defaults)
+      young_reserve = (_heap->young_generation()->max_capacity() * ShenandoahEvacReserve) / 100;
+      old_reserve = MAX2((_heap->old_generation()->max_capacity() * ShenandoahOldEvacReserve) / 100,
+                         ShenandoahOldCompactionReserve * ShenandoahHeapRegion::region_size_bytes());
+    }
   }
-
+  reserve_regions(young_reserve, old_reserve);
```

Comment:
So I see now that this does a further iteration (below) over the regions until a specific size criteria is met, albeit iterating over the regions in the opposite order. While I have not understood the high level idea of the two loops, one thing to consider if whether the two iterations, one forward and one backward are needed, or if they could somehow be fused into a single iteration. Just tossing it out there as a question that you might be able to answer immediately, probably in the negative.

A brief couple of lines describing the idea of the two loops above and below would be useful.

Reply:
I've added comments to explain rationale for the backward loop.